x-pack/metricbeat/module/openai/usage/usage.go

```diff
@@ -0,0 +1,377 @@
+func (m *MetricSet) fetchDateRange(startDate, endDate time.Time, httpClient *RLHTTPClient) error {
+	for _, apiKey := range m.config.APIKeys {
+		// stateKey using stateManager's key prefix and hashing apiKey
+		stateKey := m.stateManager.GetStateKey(apiKey.Key)
+
+		lastProcessedDate, err := m.stateManager.GetLastProcessedDate(apiKey.Key)
+		if err == nil {
+			// We have previous state, adjust start date
+			startDate = lastProcessedDate.AddDate(0, 0, 1)
+			if startDate.After(endDate) {
+				continue
+			}
+		}
+
+		for d := startDate; !d.After(endDate); d = d.AddDate(0, 0, 1) {
+			dateStr := d.Format("2006-01-02")
+			if err := m.fetchSingleDay(dateStr, apiKey.Key, httpClient); err != nil {
+				m.logger.Errorf("Error fetching data for date %s: %v", dateStr, err)
+				continue
+			}
+			if err := m.stateManager.store.Put(stateKey, dateStr); err != nil {
+				m.logger.Errorf("Error storing state for API key: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// fetchSingleDay retrieves usage data for a specific date and API key.
+func (m *MetricSet) fetchSingleDay(dateStr, apiKey string, httpClient *RLHTTPClient) error {
+	req, err := m.createRequest(dateStr, apiKey)
+	if err != nil {
+		return fmt.Errorf("error creating request: %w", err)
+	}
+
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		var netErr net.Error
+		if errors.As(err, &netErr) && netErr.Timeout() {
+			return fmt.Errorf("request timed out with configured timeout: %v and error: %w", m.config.Timeout, err)
+		}
+		return fmt.Errorf("error making request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("error response from API: status=%s", resp.Status)
+	}
+
+	return m.processResponse(resp, dateStr)
+}
+
+// createRequest builds an HTTP request for the OpenAI usage API.
+func (m *MetricSet) createRequest(dateStr, apiKey string) (*http.Request, error) {
+	req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, m.config.APIURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %w", err)
+	}
+
+	q := req.URL.Query()
+	q.Add("date", dateStr)
+	req.URL.RawQuery = q.Encode()
+
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", apiKey))
+	for key, value := range m.headers {
+		req.Header.Add(key, value)
+	}
+
+	return req, nil
+}
+
+// processResponse handles the API response and processes the usage data.
+func (m *MetricSet) processResponse(resp *http.Response, dateStr string) error {
+	var usageResponse UsageResponse
+	if err := json.NewDecoder(resp.Body).Decode(&usageResponse); err != nil {
+		return fmt.Errorf("error decoding response: %w", err)
+	}
+
+	m.logger.Infof("Fetched usage metrics for date: %s", dateStr)
+
+	events := make([]mb.Event, 0, len(usageResponse.Data))
+
+	m.processUsageData(events, usageResponse.Data)
+	m.processDalleData(events, usageResponse.DalleApiData)
+	m.processWhisperData(events, usageResponse.WhisperApiData)
```

Comment:
From here looks like we are appending all the data to a single event, even if the model is not present. So the values might be 0. Could that become mis information for te user ?

Reply:
Nope. If you see the sample in the PR's description, you will see only the entries when only it used. So, OpenAI itself appends an object to the JSON array iff there's an usage. So, when we unmarshall the JSON, we take care of that.

So, if you see inside these functions, it's ranging over a loop. But if there's like 0, then it won't. Also, in `processEvents` method, events are published iff the len of slice is more than 0.